src/unity/toolkits/supervised_learning/automatic_model_creation.cpp

```diff
@@ -1,17 +1,155 @@
-#include <unity/lib/api/unity_sframe_interface.hpp>
 #include <unity/toolkits/supervised_learning/automatic_model_creation.hpp>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <unity/lib/api/unity_sframe_interface.hpp>
+#include <unity/lib/unity_sframe.hpp>
 #include <unity/toolkits/supervised_learning/boosted_trees.hpp>
 #include <unity/toolkits/supervised_learning/decision_tree.hpp>
 #include <unity/toolkits/supervised_learning/linear_regression.hpp>
 #include <unity/toolkits/supervised_learning/linear_svm.hpp>
 #include <unity/toolkits/supervised_learning/logistic_regression.hpp>
 #include <unity/toolkits/supervised_learning/random_forest.hpp>
 #include <unity/toolkits/supervised_learning/supervised_learning.hpp>
+#include <unity/toolkits/supervised_learning/supervised_learning_utils-inl.hpp>
 
 
 namespace turi {
 namespace supervised {
 
+namespace {
+
+// Threshold on the number of feature columns, used to determine the
+// applicability of certain models.
+constexpr size_t WIDE_DATA = 200;
+
+/**
+ * Compute the width of the data.
+ *
+ * \param[in] X  Input SFrame
+ * \returns width
+ *
+ * The width is the same as the num_coefficients.
+ *
+ */
+size_t compute_data_width(sframe X){
+  ml_data data;
+  data.fill(X);
+  return get_number_of_coefficients(data.metadata());
+}
+
+
+// If `field_name` is present in `model_fields`, populates `*output` with the
+// value of `model->get_value_from_state(field_name)` and returns
+// true. Otherwise, returns false.
+bool read_model_field(
+    const std::shared_ptr<supervised_learning_model_base>& model,
+    const std::vector<std::string>& model_fields,
+    const char* field_name, variant_type* output) {
+  auto iter = std::find(model_fields.begin(), model_fields.end(), field_name);
+  bool success = iter != model_fields.end();
+  if (success) {
+    *output = model->get_value_from_state(field_name);
+  }
+  return success;
+}
+
+// Returns the regressors worth trying for the given data.
+std::vector<std::shared_ptr<supervised_learning_model_base>>
+get_regression_models(std::shared_ptr<unity_sframe> X) {
+
+  std::vector<std::shared_ptr<supervised_learning_model_base>> models;
+
+  const size_t data_width = compute_data_width(*X->get_underlying_sframe());
+
+  // Always try linear regression.
+  models.push_back(std::make_shared<linear_regression>());
+
+  if (data_width < WIDE_DATA) {
+    models.push_back(std::make_shared<xgboost::boosted_trees_regression>());
+  }
+
+  return models;
+}
+
+// Returns the validation Root Mean Squared Error if it exists, otherwise
+// returns training RMSE. Throws if neither field exists.
+double get_regression_rmse(
+    const std::shared_ptr<supervised_learning_model_base>& model) {
+
+  variant_type rmse;
+  std::vector<std::string> model_fields = model->list_fields();
+  bool success = false;
```

Comment:
instead of
```
bool success = false;
if (!success) {
    success = read_model_field(model, model_fields, "validation_rmse", &rmse);
}
```
Maybe just:
`bool success = read_model_field(model, model_fields, "validation_rmse", &rmse);`